plugins/trader/portfolioManager.js

```diff
@@ -44,12 +45,43 @@ var Manager = function(conf) {
   this.currency = conf.currency;
   this.asset = conf.asset;
   this.keepAsset = 0;
+  this.tradeTimeout = 0;
+
+  if(_.isNumber(conf.tradeTimeout)) {
+    log.debug('Trade timer is active.  Trades will expire after ' + conf.tradeTimeout + ' seconds.');
+    util.minToMs
+    this.tradeTimeout = util.secToMs(conf.tradeTimeout);
+  }
 
   if(_.isNumber(conf.keepAsset)) {
     log.debug('Keep asset is active. Will try to keep at least ' + conf.keepAsset + ' ' + conf.asset);
     this.keepAsset = conf.keepAsset;
   }
 
+  // Attempt to use websockets
+  if(conf.websocket == true) {
+    var wserror = checker.cantWebsocket(conf);
+    if(wserror) {
+      log.debug(wserror);
+      this.websocket= false;
+    } else {
+      this.websocket = true;
+      var ws = function(error) {
+          log.debug('Websocket Error: ' + error);
+          log.debug('Disabling websockets due to errors. Gekko will proceed with REST API as normal.');
+          this.websocket = false;
+      }.bind(this);
+      if(_.isNumber(conf.tradeSpeed)) {
+        this.tradeSpeed = conf.tradeSpeed;
+      } else {
+        this.tradeSpeed = 100;
+      }
+      log.debug('Websockets enabled, positions will be managed more quickly via websocket data.');
+      log.debug('Trade positions will be managed at ' + this.tradeSpeed + 'ms intervals.');
+      this.exchange.startWebsocket(ws);
```

Comment:
I don't think this is really the right place for this code. Ideally the exchange shouldn't need to export knowledge that it is a websocket to start the websocket, this should pretty much be implicit to the exchange implementation. Could this be moved to the GDAx constructor?

Reply:
After working on it for a while, I have gotten sort of lost in what should go where.

Several things access the exchange object, portfolioManager, marketWatcher being the most notable, and I didn't want there to be multiple websocket streams getting opened.

My issue is that I want all of the modules which use an exchange to use the same websocket feed and corresponding arrays / objects created by it.  I got sort of stuck and I have been trying to figure it out / hoping someone with more JS expertise might take a look.

See my previous comment about some thoughts I have, specifically with respect to how the tick system works and in what ways gekko could leverage a websocket stream vs. api calls and how much of the core would have to be changed to accommodate it...

I think I got a little in over my head to be honest.

Reply:
Take at look at Yoyae's repo which uses ccxt to manage the connections to the exchanges. I think gekko would be better if it utilized ccxt as it is literally updated multiple times a day and the dev team can focus on core of the project.

Reply:
I think using CCXT in gekko would be great, but that isn't really in the scope of this PR.  This PR is attempting to make Gekko capable of using Websocket data from an exchange in a meaningful way.

Reply:
CCXT also does not support websockets.